apispec 4.0.0

apispec just released a new version containing breaking changes

and it is installed due to setup requirements:

```
REQUIRES = [
    [...]
    'apispec>=1.0.0',
]
```

Error:

```
/usr/local/lib/python3.8/dist-packages/flask_apispec/apidoc.py:92: in get_parameters
    converter = openapi.fields2parameters
E   AttributeError: 'OpenAPIConverter' object has no attribute 'fields2parameters'
```


This is from the changelog of apispec:

> Refactoring:
> 
> OpenAPIConverter.field2parameters and OpenAPIConverter.property2parameter are removed.
> OpenAPIConverter.field2parameter becomes private. (:pr:`581`)

As a temporary fix (while working to support apispec 4.x) a 

```
REQUIRES = [
    [...]
    'apispec>=1.0.0,<4',
]
```

would be helpful

--- a/.travis.yml
+++ b/.travis.yml
@@ -5,8 +5,8 @@ env:
 - MARSHMALLOW_VERSION="==3.0.0"
 - MARSHMALLOW_VERSION=""
 python:
-- '3.5'
 - '3.6'
+- '3.8'
 before_install:
 - travis_retry pip install codecov
 install:
@@ -21,7 +21,7 @@ jobs:
   include:
   - stage: PyPI Release
     if: tag IS present
-    python: "3.6"
+    python: "3.8"
     env: []
     # Override install, and script to no-ops
     before_install: true

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -1,6 +1,14 @@
 Changelog
 ---------
 
+0.11.0 (unreleased)
+*******************
+
+Features:
+
+* Support apispec>=4.0.0 (:issue:`202`). Thanks :user:`kam193`.
+  *Backwards-incompatible*: apispec<4.0.0 is no longer supported.
+
 0.10.1 (2020-10-25)
 *******************
 

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -9,6 +9,10 @@ Features:
 * Support apispec>=4.0.0 (:issue:`202`). Thanks :user:`kam193`.
   *Backwards-incompatible*: apispec<4.0.0 is no longer supported.
 
+Other changes:
+
+* *Backwards-incompatible*: Drop Python 3.5 compatibility. Only Python>=3.6 is supported.
+
 0.10.1 (2020-10-25)
 *******************
 

--- a/flask_apispec/apidoc.py
+++ b/flask_apispec/apidoc.py
@@ -72,10 +72,7 @@ class Converter:
         return None
 
     def get_parameters(self, rule, view, docs, parent=None):
-        if APISPEC_VERSION_INFO[0] < 3:
-            openapi = self.marshmallow_plugin.openapi
-        else:
-            openapi = self.marshmallow_plugin.converter
+        openapi = self.marshmallow_plugin.converter
         annotation = resolve_annotations(view, 'args', parent)
         extra_params = []
         for args in annotation.options:

--- a/flask_apispec/apidoc.py
+++ b/flask_apispec/apidoc.py
@@ -1,4 +1,5 @@
 import copy
+import functools
 
 import apispec
 from apispec.core import VALID_METHODS
@@ -15,7 +16,6 @@ APISPEC_VERSION_INFO = tuple(
 )
 
 class Converter:
-
     def __init__(self, app, spec, document_options=True):
         self.app = app
         self.spec = spec
@@ -80,23 +80,19 @@ class Converter:
         extra_params = []
         for args in annotation.options:
             schema = args.get('args', {})
-            if is_instance_or_subclass(schema, Schema):
-                converter = openapi.schema2parameters
-            elif callable(schema):
-                schema = schema(request=None)
-                if is_instance_or_subclass(schema, Schema):
-                    converter = openapi.schema2parameters
+            openapi_converter = openapi.schema2parameters
+            if not is_instance_or_subclass(schema, Schema):
+                if callable(schema):
+                    schema = schema(request=None)
                 else:
-                    converter = openapi.fields2parameters
-            else:
-                converter = openapi.fields2parameters
+                    schema = Schema.from_dict(schema)
+                    openapi_converter = functools.partial(
+                        self._convert_dict_schema, openapi_converter)
+
             options = copy.copy(args.get('kwargs', {}))
-            location = options.pop('location', None)
-            if location:
-                options['default_in'] = location
-            elif 'default_in' not in options:
-                options['default_in'] = 'body'
-            extra_params += converter(schema, **options) if args else []
+            if not options.get('location'):
+                options['location'] = 'body'
+            extra_params += openapi_converter(schema, **options) if args else []
 
         rule_params = rule_to_params(rule, docs.get('params')) or []
 
@@ -105,6 +101,33 @@ class Converter:
     def get_responses(self, view, parent=None):
         annotation = resolve_annotations(view, 'schemas', parent)
         return merge_recursive(annotation.options)
+
+    def _convert_dict_schema(self, openapi_converter, schema, location, **options):
+        """When location is 'body' and OpenApi is 2, return one param for body fields.
+
+        Otherwise return fields exactly as converted by apispec."""
+        if self.spec.openapi_version.major < 3 and location == 'body':
+            params = openapi_converter(schema, location=None, **options)
+            body_parameter = {
+                "in": "body",
+                "name": "body",
+                "required": False,
+                "schema": {
+                    "type": "object",
+                    "properties": {},
+                },
+            }
+            for param in params:
+                name = param["name"]
+                body_parameter["schema"]["properties"].update({name: param})
+                if param.get("required", False):
+                    body_parameter["schema"].setdefault("required", []).append(name)
+                del param["name"]
+                del param["in"]
+                del param["required"]
+            return [body_parameter]
+
+        return openapi_converter(schema, location=location, **options)
 
 class ViewConverter(Converter):
 

--- a/setup.py
+++ b/setup.py
@@ -48,7 +48,7 @@ setup(
     license='MIT',
     zip_safe=False,
     keywords='flask marshmallow webargs apispec',
-    python_requires=">=3.5",
+    python_requires=">=3.6",
     test_suite='tests',
     project_urls={
         'Bug Reports': 'https://github.com/jmcarp/flask-apispec/issues',

--- a/setup.py
+++ b/setup.py
@@ -6,7 +6,7 @@ REQUIRES = [
     'flask>=0.10.1',
     'marshmallow>=3.0.0',
     'webargs>=6.0.0',
-    'apispec>=1.0.0',
+    'apispec>=4.0.0',
 ]
 
 

--- a/tests/test_openapi.py
+++ b/tests/test_openapi.py
@@ -25,10 +25,7 @@ def spec(marshmallow_plugin):
 
 @pytest.fixture()
 def openapi(marshmallow_plugin):
-    if APISPEC_VERSION_INFO[0] < 3:
-        return marshmallow_plugin.openapi
-    else:
-        return marshmallow_plugin.converter
+    return marshmallow_plugin.converter
 
 def ref_path(spec):
     if spec.openapi_version.version[0] < 3:

--- a/tests/test_openapi.py
+++ b/tests/test_openapi.py
@@ -1,4 +1,5 @@
 import pytest
+from unittest import mock
 from apispec import APISpec
 from apispec.ext.marshmallow import MarshmallowPlugin
 from marshmallow import fields, Schema
@@ -7,7 +8,7 @@ from flask import make_response
 from flask_apispec.paths import rule_to_params
 from flask_apispec.views import MethodResource
 from flask_apispec import doc, use_kwargs, marshal_with
-from flask_apispec.apidoc import APISPEC_VERSION_INFO, ViewConverter, ResourceConverter
+from flask_apispec.apidoc import APISPEC_VERSION_INFO, Converter, ViewConverter, ResourceConverter
 
 @pytest.fixture()
 def marshmallow_plugin():
@@ -113,8 +114,8 @@ class TestArgSchema:
         params = path['get']['parameters']
         rule = app.url_map._rules_by_endpoint['get_band'][0]
         expected = (
-            openapi.fields2parameters(
-                {'name': fields.Str()}, default_in='query') +
+            openapi.schema2parameters(
+                Schema.from_dict({'name': fields.Str()}), location='query') +
             rule_to_params(rule)
         )
         assert params == expected
@@ -184,8 +185,7 @@ class TestResourceView:
         params = path['get']['parameters']
         rule = app.url_map._rules_by_endpoint['band'][0]
         expected = (
-            openapi.fields2parameters(
-                {'name': fields.Str()}, default_in='query') +
+            [{'in': 'query', 'name': 'name', 'required': False, 'type': 'string'}] +
             rule_to_params(rule)
         )
         assert params == expected
@@ -242,7 +242,6 @@ class TestMultipleLocations:
         )
         assert params == expected
 
-
 class TestGetFieldsNoLocationProvided:
 
     @pytest.fixture
@@ -277,6 +276,33 @@ class TestGetFieldsNoLocationProvided:
             },
         } in params
 
+class TestGetFieldsBodyLocation(TestGetFieldsNoLocationProvided):
+
+    @pytest.fixture
+    def function_view(self, app):
+        @app.route('/bands/<int:band_id>/')
+        @use_kwargs({'name': fields.Str(required=True), 'address': fields.Str(), 'email': fields.Str(required=True)})
+        def get_band(**kwargs):
+            return kwargs
+
+        return get_band
+
+    def test_params(self, app, path):
+        params = path['get']['parameters']
+        assert {
+            'in': 'body',
+            'name': 'body',
+            'required': False,
+            'schema': {
+                'properties': {
+                    'address': {'type': 'string'},
+                    'name': {'type': 'string'},
+                    'email': {'type': 'string'},
+                },
+                'required': ["name", "email"],
+                'type': 'object',
+            },
+        } in params
 
 class TestSchemaNoLocationProvided:
 

--- a/tests/test_openapi.py
+++ b/tests/test_openapi.py
@@ -1,5 +1,4 @@
 import pytest
-from unittest import mock
 from apispec import APISpec
 from apispec.ext.marshmallow import MarshmallowPlugin
 from marshmallow import fields, Schema
@@ -8,7 +7,7 @@ from flask import make_response
 from flask_apispec.paths import rule_to_params
 from flask_apispec.views import MethodResource
 from flask_apispec import doc, use_kwargs, marshal_with
-from flask_apispec.apidoc import APISPEC_VERSION_INFO, Converter, ViewConverter, ResourceConverter
+from flask_apispec.apidoc import ViewConverter, ResourceConverter
 
 @pytest.fixture()
 def marshmallow_plugin():

--- a/tox.ini
+++ b/tox.ini
@@ -1,5 +1,5 @@
 [tox]
-envlist=py36,py38,pypy
+envlist=py35,py36,py37,py38,py39,pypy
 [testenv]
 deps=
     -rdev-requirements.txt

--- a/tox.ini
+++ b/tox.ini
@@ -1,5 +1,5 @@
 [tox]
-envlist=py27,py35,py36,pypy
+envlist=py36,py38,pypy
 [testenv]
 deps=
     -rdev-requirements.txt

